// content-scripts-register-polyfill/index.js
"use strict";
/// <reference path="./globals.d.ts" />
Object.defineProperty(exports, "__esModule", { value: true });
const webext_patterns_1 = require("webext-patterns");
// @ts-expect-error
async function p(fn, ...args) {
    return new Promise((resolve, reject) => {
        // @ts-expect-error
        fn(...args, result => {
            if (chrome.runtime.lastError) {
                reject(chrome.runtime.lastError);
            }
            else {
                resolve(result);
            }
        });
    });
}
async function isOriginPermitted(url) {
    return p(chrome.permissions.contains, {
        origins: [new URL(url).origin + '/*']
    });
}
async function wasPreviouslyLoaded(tabId, loadCheck) {
    const result = await p(chrome.tabs.executeScript, tabId, {
        code: loadCheck,
        runAt: 'document_start'
    });
    return result === null || result === void 0 ? void 0 : result[0];
}
if (typeof chrome === 'object' && !chrome.contentScripts) {
    chrome.contentScripts = {
        // The callback is only used by webextension-polyfill
        async register(contentScriptOptions, callback) {
            const { js = [], css = [], allFrames, matchAboutBlank, matches, runAt } = contentScriptOptions;
            // Injectable code; it sets a `true` property on `document` with the hash of the files as key.
            const loadCheck = `document[${JSON.stringify(JSON.stringify({ js, css }))}]`;
            const matchesRegex = webext_patterns_1.patternToRegex(...matches);
            const listener = async (tabId, { status }) => {
                if (status !== 'loading') {
                    return;
                }
                const { url } = await p(chrome.tabs.get, tabId);
                if (!url || // No URL = no permission;
                    !matchesRegex.test(url) || // Manual `matches` glob matching
                    !await isOriginPermitted(url) || // Permissions check
                    await wasPreviouslyLoaded(tabId, loadCheck) // Double-injection avoidance
                ) {
                    return;
                }
                for (const file of css) {
                    chrome.tabs.insertCSS(tabId, {
                        ...file,
                        matchAboutBlank,
                        allFrames,
                        runAt: runAt !== null && runAt !== void 0 ? runAt : 'document_start' // CSS should prefer `document_start` when unspecified
                    });
                }
                for (const file of js) {
                    chrome.tabs.executeScript(tabId, {
                        ...file,
                        matchAboutBlank,
                        allFrames,
                        runAt
                    });
                }
                // Mark as loaded
                chrome.tabs.executeScript(tabId, {
                    code: `${loadCheck} = true`,
                    runAt: 'document_start',
                    allFrames
                });
            };
            chrome.tabs.onUpdated.addListener(listener);
            const registeredContentScript = {
                async unregister() {
                    return p(chrome.tabs.onUpdated.removeListener.bind(chrome.tabs.onUpdated), listener);
                }
            };
            if (typeof callback === 'function') {
                callback(registeredContentScript);
            }
            return Promise.resolve(registeredContentScript);
        }
    };
}
